passes/techmap/clockgate.cc

```diff
@@ -37,6 +39,161 @@ ClockGateCell icg_from_arg(std::string& name, std::string& str) {
+				icg_kind.c_str(), cell_name.c_str());
+		auto internal = kind_tags[0];
+		if (internal != "latch") {
+			// TODO Is this expected behavior?
+			log_warning("Skipping ICG cell %s - not latch-based\n", cell_name.c_str());
+			continue;
+		}
+		auto clk_pol_tag = kind_tags[1];
+		bool clk_pol;
+		if (clk_pol_tag == "posedge") {
+			clk_pol = true;
+		} else if (clk_pol_tag == "negedge") {
+			clk_pol = false;
+		} else {
+			log_error("Malformed liberty file - invalid clock_gating_integrated_cell value %s in cell %s\n",
+				icg_kind.c_str(), cell_name.c_str());
+			continue;
+		}
+		log_debug("maybe valid icg: %s\n", cell_name.c_str());
+		ClockGateCell icg_interface;
+		icg_interface.name = RTLIL::escape_id(cell_name);
+
+		for (auto pin : cell->children) {
+			if (pin->id != "pin" || pin->args.size() != 1)
+				continue;
+
+			log("Check pin %s\n", pin->args[0].c_str());
+			if (auto clk = pin->find("clock_gate_clock_pin")) {
+				if (!icg_interface.clk_in_pin.empty())
+					log_error("Malformed liberty file - multiple clock_gate_clock_pin in cell %s\n",
+						cell_name.c_str());
+				else
+					icg_interface.clk_in_pin = RTLIL::escape_id(pin->args[0]);
+			} else if (auto gclk = pin->find("clock_gate_out_pin")) {
+				if (!icg_interface.clk_out_pin.empty())
+					log_error("Malformed liberty file - multiple clock_gate_out_pin in cell %s\n",
+						cell_name.c_str());
+				else
+					icg_interface.clk_out_pin = RTLIL::escape_id(pin->args[0]);
+			} else if (auto en = pin->find("clock_gate_enable_pin")) {
+				if (!icg_interface.ce_pin.empty())
+					log_error("Malformed liberty file - multiple clock_gate_enable_pin in cell %s\n",
+						cell_name.c_str());
+				else
+					icg_interface.ce_pin = RTLIL::escape_id(pin->args[0]);
+			} else if (auto se = pin->find("clock_gate_test_pin")) {
+				icg_interface.tie_lo_pins.push_back(RTLIL::escape_id(pin->args[0]));
+			} else {
+				const LibertyAst *dir = pin->find("direction");
+				if (dir->value == "internal")
+					continue;
+
+				log_error("Malformed liberty file - extra pin %s in cell %s\n",
+					pin->args[0].c_str(), cell_name.c_str());
+			}
+		}
+
+		if (icg_interface.clk_in_pin.empty())
+			log_error("Malformed liberty file - missing clock_gate_clock_pin in cell %s",
+				cell_name.c_str());
+		if (icg_interface.clk_out_pin.empty())
+			log_error("Malformed liberty file - missing clock_gate_out_pin in cell %s",
+				cell_name.c_str());
+		if (icg_interface.ce_pin.empty())
+			log_error("Malformed liberty file - missing clock_gate_enable_pin in cell %s",
+				cell_name.c_str());
+
+		double area = 0;
+		const LibertyAst *ar = cell->find("area");
+		if (ar != nullptr && !ar->value.empty())
+			area = atof(ar->value.c_str());
+
+		std::optional<ICGRankable>& icg_to_beat = clk_pol ? best_pos : best_neg;
+
+		bool winning = false;
+		if (icg_to_beat) {
+			log_debug("ties: %zu ? %zu\n", icg_to_beat->icg.tie_lo_pins.size(),
```

Comment:
This will simplify the code around here to
```
if (!icg_to_beat ||
    std::make_pair(...) < std::make_pair(...))
  icg_to_beat.emplace(new_icg);

Reply:
Oh neat, this didn't occur to me, thanks